InDrops data can be processed by zUMIs when generated by the "v3" protocol. 
Thus, after bcl2fastq you will get four fastq files from your Illumina run.

The Reads correspond to:
* R1 = cDNA
* R2 = first half of gel barcode (8 nt)
* R3 = library barcode (6 nt)
* R4 = second half of gel barcode + UMI + polyT (8 + 6 + 8)

You need to swith on the InDrops mode with the -Y flag. (eg. "-Y yes")
For the -c flag, please use the combined barcode lengths (eg. -c 1-22).
Next, set the UMI base-ranges to start after the cell barcode (eg. -m 23-28).

Provide the following additional settings in InDrops:
### InDrops parameters ###
	-Y  <InDrops>		 : Do you have InDrops data? yes/no Default: no.
	-F  <gel-barcode2 fastq> : Provide the second half of gel barcode + UMI read <-F> here. Default: NA.
	-L  <library barcode fastq> : Provide the library barcode read here. Default: NA


Here is an InDrops example:

```
bash zUMIs-master.sh -f barcode1read.R2.fastq -L librarybc.R3.fastq -F barcode2UMIread.R4.fastq -r cdnaread.R1.fastq -n test_InDrops -g hg38_STAR5idx_noGTF/  -a Homo_sapiens.GRCh38.84.gtf -p 8 -c 1-22 -m 23-28 -l 57 -q 17 -Q 17 -Y yes
```